### ISSUE TITLE
Think through how we want to track provenance in the PDC

### ISSUE THREAD

**Issue by slifty** [PhilanthropyDataCommons/service#1083]
Though I think I have a general sense of user stories, we should spend a moment codifying them and thinking about what might be missed.

Once that's done, the key questions will be around how to store it appropriately.  Once that's done we'll make separate issues for actually implementing it and making it available via the API.

**Comment by slifty** [comment]
Some related notes [from our meeting](https://pad.opentechstrategies.com/p/pdc-dev-2024-07-08):

- Permissioning.  "If a piece of data came in from Funder Foo, then I'm okay with it being shared.  If it came from ${OTHER_SOURCE}, then I'm not."
- Especially important for, say, funder collaboratives.
- Data reconciliation and collision resolution
- Actually answering users' questions about where something came from.

**Comment by bickelj** [comment]
TLDR: for each entity, add one field for those who create data (e.g. data platform provider) and one (optional) field for true owners of data (e.g. changemaker/applicant/organization).

If we separate provenance from ownership (we should), and if any endpoint that allows modification (i.e. `POST`, `PUT`, etc.) requires a JWT (it should), the simple thing for provenance is to store the JWT in that entity that was created. If we store the JWT we can invalidate it for replay by stripping the signature like the current logs have it.

Storing the ID of the one that posted data doesn't solve ownership at all, which is more challenging. I think "if I created it, I have maximum permissions on it" is a good rule. That does not exclude maximum permissions from the true owner. A separate field for true owner is needed. Figuring out what to put in that field and when to fill it is not clear. If the true owner exists in our auth system, I suppose we could require the one who posts to PDC to include our ID for that owner. If the true owner does not exist in our system, do we auto-create it? Do we require the "owner" to actually be a group or an organization rather than an individual? Perhaps that is an approach: the "I created this record" field is an individual Keycloak account while the "I own this record" field is a Keycloak group. That raises the question as to how those Keycloak groups are created and managed as well as deferral of assigning true ownership until that group exists.

The approach described should solve most of the use cases with the word "my" or "mine" in it, but not all of them.

**Comment by slifty** [comment]
I think that we should create a new entity of `source` which represents the source of a given set of data.

I think it would be reasonable for us to model provenance at the object that is closest to the act of insertion -- not the data itself.

What I mean by this based on our current use cases is:

1. What agent + account initiated a proposal version upload, and what was the source.
2. What agent + account initiated an organizational data upload, and what was the source.

This source + agent + account association would be made regardless of whether the data came from a direct API call, or from a bulk upload / batch operation of some kind.

This is not directly related to this, but this model would eventually allow us to signal things like source type (GMS, 3rd party data provider, direct entry, etc), as well as source reliability / quality ratings which could inform things like #1087.

Implementation of this would entail:

1. Creation of a new `DataSource` entity in the schema, as well as CRUD endpoints.
2. Addition of a `dataSourceId` field in `ProposalVersion`.
3. Eventual incorporation of `dataSourceId` in our implementation of #1088 

I agree that this concept should be completely separated from business logic related to ownership and permissions.  For example, knowing that an account created a given piece of data does NOT mean that account has any ownership over that data in the long term.  That said, we'll still want to know who added the data.

**Comment by bickelj** [comment]
I'm OK with the approach @slifty mentioned. I think it's compatible with what I suggested above: a field for data provenance and also a field for data ownership. My main objection is the new entity name must not be the overloaded term `DataSource`. Maybe `DataOrigin`, `DataProvenance` would work. Perhaps more simply `Provenance`.

I advocate for the "if I created it, I am the default owner until I specify otherwise" model, though. I think the Keycloak account that created the row is an essential part of the provenance objects. @slifty perhaps that's what you mean by `agent`?

**Comment by slifty** [comment]
Oh snap `Provenance` is great.

Also I totally am down for that default re ownership -- I was mostly suggesting we might want to decouple the ownership decisions from the provenance tracking itself (I imagine we might want ownership to be more complicated / involve an access list model / not be a column in the data table so that it can support things like multiple owners, removal of access without removal of provenance, etc, etc)

**Comment by slifty** [comment]
We had some discussion about how `provenance` might not be the right word for this just because it indicates both the actor (account) as well as the "source"

Origin might not be right because it might signal "ultimate origin" rather than simply "which data source this information came from" -- then again maybe it's fine!

I'll take this one on and think through naming as part of implementation.

**Comment by slifty** [comment]
I just realized that the issue is functionally closed!  I'll make an issue to capture the actual work / reference this issue as the........ ORIGIN of the implementation plan.
